hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotRenameRequest.java

```diff
@@ -0,0 +1,243 @@
+      }
+
+      switch (snapshotOldInfo.getSnapshotStatus()) {
+      case SNAPSHOT_DELETED:
+        throw new OMException("Snapshot is already deleted. "
+            + "Pending reclamation.", FILE_NOT_FOUND);
+      case SNAPSHOT_ACTIVE:
+        break;
+      default:
+          // Unknown snapshot non-active state
+        throw new OMException("Snapshot exists but no longer in active state",
+            FILE_NOT_FOUND);
+      }
+
+      snapshotOldInfo.setName(snapshotNewName);
+
+      omMetadataManager.getSnapshotInfoTable().addCacheEntry(
+          new CacheKey<>(snapshotOldTableKey),
+          CacheValue.get(termIndex.getIndex()));
+
+      omMetadataManager.getSnapshotInfoTable().addCacheEntry(
+          new CacheKey<>(snapshotNewTableKey),
+          CacheValue.get(termIndex.getIndex(), snapshotOldInfo));
+
+      updateSnapshotInChainManager(omMetadataManager, snapshotOldInfo);
+
+      ozoneManager.getOmSnapshotManager().getSnapshotCache().invalidate(snapshotOldTableKey);
+
+      omResponse.setRenameSnapshotResponse(
+          OzoneManagerProtocolProtos.RenameSnapshotResponse.newBuilder()
+              .setSnapshotInfo(snapshotOldInfo.getProtobuf()));
+      omClientResponse = new OMSnapshotRenameResponse(
+          omResponse.build(), snapshotOldTableKey, snapshotNewTableKey, snapshotOldInfo);
+
+    } catch (IOException | InvalidPathException ex) {
+      exception = ex;
+      omClientResponse = new OMSnapshotRenameResponse(
+          createErrorOMResponse(omResponse, exception));
+    } finally {
+      if (acquiredSnapshotNewLock) {
+        mergeOmLockDetails(omMetadataManager.getLock().releaseWriteLock(SNAPSHOT_LOCK, volumeName,
+                                                     bucketName, snapshotNewName));
+      }
+      if (acquiredSnapshotOldLock) {
+        mergeOmLockDetails(omMetadataManager.getLock().releaseWriteLock(SNAPSHOT_LOCK, volumeName,
+                                                     bucketName, snapshotOldName));
+      }
+      if (acquiredBucketLock) {
+        mergeOmLockDetails(omMetadataManager.getLock().releaseWriteLock(BUCKET_LOCK, volumeName,
+                                                     bucketName));
+      }
+      if (omClientResponse != null) {
+        omClientResponse.setOmLockDetails(getOmLockDetails());
+      }
+    }
+
+    if (snapshotOldInfo == null) {
+      // Dummy SnapshotInfo for logging and audit logging when erred
+      snapshotOldInfo = SnapshotInfo.newInstance(volumeName, bucketName,
+                                              snapshotOldName, null, Time.now());
+    }
+
+    // Perform audit logging outside the lock
+    auditLog(auditLogger, buildAuditMessage(OMAction.RENAME_SNAPSHOT,
+                                            snapshotOldInfo.toAuditMap(), exception, userInfo));
+    return omClientResponse;
+  }
+
+  private void updateSnapshotInChainManager(OmMetadataManagerImpl omMetadataManager, SnapshotInfo snapshotInfo) {
```

Comment:
nit: This doesn't need to be a separate function. Please directly call `snapshotChainManager.updateSnapshot(snapshotInfo);` since `updateSnapshot()` is synchronized operation.